File: artifacts/api-server/api/views/ingest.py
import csv
import io
import re
from datetime import datetime, date, timezone as dt_timezone
from rest_framework.views import APIView
from rest_framework.response import Response
from rest_framework import status
from django.utils import timezone
from ..models import Client, IngestionBatch, EmissionsRecord
from ..serializers import IngestionBatchSerializer

FUEL_FACTORS = {
    "diesel":       {"factor": 2.688, "scope": "scope1"},
    "petrol":       {"factor": 2.315, "scope": "scope1"},
    "gasoline":     {"factor": 2.315, "scope": "scope1"},
    "lng":          {"factor": 1.164, "scope": "scope1"},
    "lpg":          {"factor": 1.555, "scope": "scope1"},
    "fluessiggas":  {"factor": 1.555, "scope": "scope1"},
    "natural gas":  {"factor": 2.034, "scope": "scope1"},
    "erdgas":       {"factor": 2.034, "scope": "scope1"},
    "benzin":       {"factor": 2.315, "scope": "scope1"},
}

ELECTRICITY_FACTOR = 0.207

FLIGHT_FACTORS = {"short": 0.255, "medium": 0.195, "long": 0.150}
HOTEL_FACTOR = 31.0
GROUND_FACTOR = 0.170

AIRPORT_DISTANCES = {
    "LHR-JFK": 5540, "JFK-LHR": 5540,
    "LHR-CDG": 340,  "CDG-LHR": 340,
    "LHR-DXB": 5500, "DXB-LHR": 5500,
    "JFK-LAX": 3980, "LAX-JFK": 3980,
    "BOM-DEL": 1150, "DEL-BOM": 1150,
    "SIN-SYD": 6300, "SYD-SIN": 6300,
    "LHR-BOM": 7190, "BOM-LHR": 7190,
    "LHR-SIN": 10840,"SIN-LHR": 10840,
}


def parse_sap_date(raw: str) -> date:
    raw = raw.strip()
    if re.match(r"^\d{8}$", raw):
        return datetime.strptime(raw, "%Y%m%d").date()
    if re.match(r"^\d{2}\.\d{2}\.\d{4}$", raw):
        return datetime.strptime(raw, "%d.%m.%Y").date()
    try:
        return datetime.fromisoformat(raw).date()
    except Exception:
        return date.today()


def parse_date(raw: str) -> date:
    raw = raw.strip()
    for fmt in ("%Y-%m-%d", "%d/%m/%Y", "%m/%d/%Y", "%d-%m-%Y", "%Y%m%d"):
        try:
            return datetime.strptime(raw, fmt).date()
        except ValueError:
            continue
    return date.today()


def get_flight_factor(km: float) -> float:
    if km < 1500:
        return FLIGHT_FACTORS["short"]
    if km < 3500:
        return FLIGHT_FACTORS["medium"]
    return FLIGHT_FACTORS["long"]


def check_suspicious(co2e_kg: float, amount: float):
    if co2e_kg > 100_000:
        return True, f"CO₂e value unusually high: {co2e_kg:.0f} kg"
    if amount <= 0:
        return True, "Activity amount is zero or negative"
    return False, None


def normalize_keys(row: dict) -> dict:
    return {k.lower().strip().replace(" ", "_").replace("-", "_"): str(v).strip() for k, v in row.items()}


def parse_csv(content: str) -> list[dict]:
    for delimiter in [",", ";", "\t"]:
        try:
            reader = csv.DictReader(io.StringIO(content), delimiter=delimiter)
            rows = list(reader)
            if rows and len(rows[0]) > 1:
                return rows
        except Exception:
            continue
    return []


class IngestSAPView(APIView):
    def post(self, request):
        client_id = request.data.get("clientId")
        file = request.FILES.get("file")

        if not client_id:
            return Response({"error": "clientId is required"}, status=status.HTTP_400_BAD_REQUEST)
        if not file:
            return Response({"error": "file is required"}, status=status.HTTP_400_BAD_REQUEST)

        try:
            client = Client.objects.get(pk=client_id)
        except Client.DoesNotExist:
            return Response({"error": "Client not found"}, status=status.HTTP_400_BAD_REQUEST)

        batch = IngestionBatch.objects.create(
            client=client, source_type="sap", filename=file.name, status="processing"
        )

        row_count = 0
        error_count = 0
        errors = []

        try:
            content = file.read().decode("utf-8", errors="replace")
            rows = parse_csv(content)

            for raw_row in rows:
                row_count += 1
                row = normalize_keys(raw_row)

                qty_raw = row.get("menge") or row.get("quantity") or row.get("qty") or row.get("amount", "")
                unit = (row.get("meins") or row.get("unit") or row.get("uom", "L")).lower()
                material = (row.get("matnr") or row.get("material") or row.get("material_description", "diesel")).lower()
                plant = row.get("werks") or row.get("plant", "")
                date_raw = row.get("budat") or row.get("posting_date") or row.get("date", "")
                company_code = row.get("bukrs") or row.get("company_code", "")

                try:
                    qty = float(qty_raw.replace(",", "."))
                except (ValueError, AttributeError):
                    error_count += 1
                    errors.append(f"Row {row_count}: invalid quantity '{qty_raw}'")
                    continue

                activity_date = parse_sap_date(date_raw)

                fuel_type = "diesel"
                for key in FUEL_FACTORS:
                    if key in material:
                        fuel_type = key
                        break
                factor_info = FUEL_FACTORS.get(fuel_type, FUEL_FACTORS["diesel"])

                litres = qty
                if unit in ("gal", "gallon"):
                    litres = qty * 3.78541
                elif unit in ("m3", "cbm"):
                    litres = qty * 1000

                co2e = litres * factor_info["factor"]
                suspicious, reason = check_suspicious(co2e, qty)

                EmissionsRecord.objects.create(
                    batch=batch,
                    client=client,
                    scope=factor_info["scope"],
                    category=fuel_type,
                    source_type="sap",
                    raw_data=raw_row,
                    activity_amount=litres,
                    activity_unit="litres",
                    normalized_co2e_kg=co2e,
                    activity_date=activity_date,
                    location=plant or company_code or None,
                    description=f"{material} combustion — plant {plant} ({company_code})",
                    is_suspicious=suspicious,
                    suspicious_reason=reason,
                )

        except Exception as e:
            error_count += 1
            errors.append(f"Failed to parse file: {e}")

        batch.status = "failed" if (error_count == row_count and row_count > 0) else "completed"
        batch.row_count = row_count
        batch.error_count = error_count
        batch.error_details = "\n".join(errors[:10]) if errors else None
        batch.completed_at = timezone.now()
        batch.save()

        return Response(IngestionBatchSerializer(batch).data, status=status.HTTP_201_CREATED)


class IngestUtilityView(APIView):
    def post(self, request):
        client_id = request.data.get("clientId")
        file = request.FILES.get("file")

        if not client_id:
            return Response({"error": "clientId is required"}, status=status.HTTP_400_BAD_REQUEST)
        if not file:
            return Response({"error": "file is required"}, status=status.HTTP_400_BAD_REQUEST)

        try:
            client = Client.objects.get(pk=client_id)
        except Client.DoesNotExist:
            return Response({"error": "Client not found"}, status=status.HTTP_400_BAD_REQUEST)

        batch = IngestionBatch.objects.create(
            client=client, source_type="utility", filename=file.name, status="processing"
        )

        row_count = 0
        error_count = 0
        errors = []

        try:
            content = file.read().decode("utf-8", errors="replace")
            rows = parse_csv(content)

            for raw_row in rows:
                row_count += 1
                row = normalize_keys(raw_row)

                kwh_raw = (row.get("consumption_kwh") or row.get("kwh") or
                           row.get("usage_kwh") or row.get("consumption", ""))
                unit = (row.get("unit") or row.get("uom", "kwh")).lower().replace(" ", "")
                period_start = (row.get("period_start") or row.get("start_date") or
                                row.get("billing_start") or row.get("date", ""))
                meter_id = row.get("meter_id") or row.get("meter") or row.get("account_number", "")
                site = (row.get("location") or row.get("site") or
                        row.get("facility") or row.get("address", ""))

                try:
                    kwh = float(kwh_raw.replace(",", ""))
                except (ValueError, AttributeError):
                    error_count += 1
                    errors.append(f"Row {row_count}: invalid consumption '{kwh_raw}'")
                    continue

                if unit == "mwh":
                    kwh *= 1000
                elif unit == "gwh":
                    kwh *= 1_000_000

                activity_date = parse_date(period_start)
                co2e = kwh * ELECTRICITY_FACTOR
                suspicious, reason = check_suspicious(co2e, kwh)

                EmissionsRecord.objects.create(
                    batch=batch,
                    client=client,
                    scope="scope2",
                    category="electricity",
                    source_type="utility",
                    raw_data=raw_row,
                    activity_amount=kwh,
                    activity_unit="kWh",
                    normalized_co2e_kg=co2e,
                    activity_date=activity_date,
                    location=site or meter_id or None,
                    description=f"Electricity — meter {meter_id}" + (f" at {site}" if site else ""),
                    is_suspicious=suspicious,
                    suspicious_reason=reason,
                )

        except Exception as e:
            error_count += 1
            errors.append(f"Failed to parse file: {e}")

        batch.status = "failed" if (error_count == row_count and row_count > 0) else "completed"
        batch.row_count = row_count
        batch.error_count = error_count
        batch.error_details = "\n".join(errors[:10]) if errors else None
        batch.completed_at = timezone.now()
        batch.save()

        return Response(IngestionBatchSerializer(batch).data, status=status.HTTP_201_CREATED)


class IngestTravelView(APIView):
    def post(self, request):
        client_id = request.data.get("clientId")
        file = request.FILES.get("file")

        if not client_id:
            return Response({"error": "clientId is required"}, status=status.HTTP_400_BAD_REQUEST)
        if not file:
            return Response({"error": "file is required"}, status=status.HTTP_400_BAD_REQUEST)

        try:
            client = Client.objects.get(pk=client_id)
        except Client.DoesNotExist:
            return Response({"error": "Client not found"}, status=status.HTTP_400_BAD_REQUEST)

        batch = IngestionBatch.objects.create(
            client=client, source_type="travel", filename=file.name, status="processing"
        )

        row_count = 0
        error_count = 0
        errors = []

        try:
            content = file.read().decode("utf-8", errors="replace")
            rows = parse_csv(content)

            for raw_row in rows:
                row_count += 1
                row = normalize_keys(raw_row)

                trip_type = (row.get("trip_type") or row.get("type") or row.get("category", "flight")).lower()
                date_raw = (row.get("date") or row.get("departure_date") or
                            row.get("travel_date") or row.get("check_in", ""))
                traveler = row.get("traveler") or row.get("employee") or row.get("name", "")
                department = row.get("department") or row.get("cost_center") or row.get("team", "")

                activity_date = parse_date(date_raw)
                co2e = 0.0
                amount = 0.0
                unit = ""
                category = "flight"
                description = ""

                if "flight" in trip_type or "air" in trip_type:
                    category = "flight"
                    origin = row.get("origin") or row.get("departure") or row.get("from", "")
                    dest = row.get("destination") or row.get("arrival") or row.get("to", "")
                    dist_raw = row.get("distance_km") or row.get("distance") or row.get("km", "")
                    pax = int(row.get("passengers") or row.get("pax") or "1") if (row.get("passengers") or "1").isdigit() else 1

                    try:
                        dist_km = float(dist_raw)
                        if dist_km <= 0:
                            raise ValueError
                    except (ValueError, TypeError):
                        key = f"{origin.upper()}-{dest.upper()}"
                        dist_km = AIRPORT_DISTANCES.get(key, 2000)

                    factor = get_flight_factor(dist_km)
                    co2e = dist_km * factor * pax
                    amount = dist_km
                    unit = "km"
                    description = f"Flight {origin}→{dest} ({dist_km:.0f} km, {pax} pax) — {traveler}"

                elif "hotel" in trip_type or "accommodation" in trip_type or "lodg" in trip_type:
                    category = "hotel"
                    nights_raw = row.get("nights") or row.get("duration_nights") or "1"
                    try:
                        nights = float(nights_raw)
                    except (ValueError, TypeError):
                        nights = 1
                    co2e = nights * HOTEL_FACTOR
                    amount = nights
                    unit = "nights"
                    loc = row.get("hotel") or row.get("property") or row.get("destination") or row.get("city", "")
                    description = f"Hotel {loc} ({nights:.0f} nights) — {traveler}"

                else:
                    category = "ground_transport"
                    dist_raw = row.get("distance_km") or row.get("distance") or "50"
                    try:
                        dist_km = float(dist_raw)
                    except (ValueError, TypeError):
                        dist_km = 50
                    co2e = dist_km * GROUND_FACTOR
                    amount = dist_km
                    unit = "km"
                    frm = row.get("origin") or row.get("from") or row.get("pickup", "")
                    to = row.get("destination") or row.get("to") or row.get("dropoff", "")
                    description = f"Ground transport {frm}→{to} ({dist_km:.0f} km) — {traveler}"

                if department:
                    description += f" [{department}]"

                suspicious, reason = check_suspicious(co2e, amount)

                EmissionsRecord.objects.create(
                    batch=batch,
                    client=client,
                    scope="scope3",
                    category=category,
                    source_type="travel",
                    raw_data=raw_row,
                    activity_amount=amount,
                    activity_unit=unit,
                    normalized_co2e_kg=co2e,
                    activity_date=activity_date,
                    location=row.get("origin") or row.get("city") or None,
                    description=description,
                    is_suspicious=suspicious,
                    suspicious_reason=reason,
                )

        except Exception as e:
            error_count += 1
            errors.append(f"Failed to parse file: {e}")

        batch.status = "failed" if (error_count == row_count and row_count > 0) else "completed"
        batch.row_count = row_count
        batch.error_count = error_count
        batch.error_details = "\n".join(errors[:10]) if errors else None
        batch.completed_at = timezone.now()
        batch.save()

        return Response(IngestionBatchSerializer(batch).data, status=status.HTTP_201_CREATED)

File: artifacts/api-server/src/routes/ingest.ts
import { Router, type IRouter } from "express";
import multer from "multer";
import { parse } from "csv-parse/sync";
import { db, ingestionBatchesTable, emissionsRecordsTable, clientsTable } from "@workspace/db";
import { eq } from "drizzle-orm";
import { logger } from "../lib/logger";

const router: IRouter = Router();
const upload = multer({ storage: multer.memoryStorage(), limits: { fileSize: 10 * 1024 * 1024 } });

// ────────────────────────────────────────────────────────────
// Emission factors (kg CO2e per unit) — sourced from IPCC AR6
// and UK DESNZ/DEFRA 2023 conversion factors
// ────────────────────────────────────────────────────────────
const FUEL_FACTORS: Record<string, { kgCo2ePerLitre: number; scope: "scope1" }> = {
  diesel:    { kgCo2ePerLitre: 2.688, scope: "scope1" },
  petrol:    { kgCo2ePerLitre: 2.315, scope: "scope1" },
  gasoline:  { kgCo2ePerLitre: 2.315, scope: "scope1" },
  lng:       { kgCo2ePerLitre: 1.164, scope: "scope1" },
  lpg:       { kgCo2ePerLitre: 1.555, scope: "scope1" },
  "natural gas": { kgCo2ePerLitre: 2.034, scope: "scope1" },
};

// UK grid: 0.207 kg CO2e / kWh (2023 DEFRA)
const ELECTRICITY_FACTOR_KG_PER_KWH = 0.207;

// Flight emission factors (kg CO2e per km per passenger, including RFI)
const FLIGHT_FACTORS: Record<string, number> = {
  short: 0.255,  // <1500 km
  medium: 0.195, // 1500-3500 km
  long: 0.150,   // >3500 km
};
const HOTEL_FACTOR_KG_PER_NIGHT = 31.0; // DEFRA 2023 business travel
const GROUND_FACTOR_KG_PER_KM = 0.14;  // avg taxi/rental

// Airport distance lookup (km, great-circle, most common pairs)
const AIRPORT_DISTANCES: Record<string, number> = {
  "LHR-JFK": 5540, "JFK-LHR": 5540,
  "LHR-CDG": 340,  "CDG-LHR": 340,
  "LHR-DXB": 5500, "DXB-LHR": 5500,
  "JFK-LAX": 3980, "LAX-JFK": 3980,
  "BOM-DEL": 1150, "DEL-BOM": 1150,
  "SIN-SYD": 6300, "SYD-SIN": 6300,
};

function getFlightDistance(origin: string, dest: string): number {
  const key = `${origin.toUpperCase()}-${dest.toUpperCase()}`;
  return AIRPORT_DISTANCES[key] ?? 2000; // default 2000 km if pair unknown
}

function getFlightFactor(distanceKm: number): number {
  if (distanceKm < 1500) return FLIGHT_FACTORS.short!;
  if (distanceKm < 3500) return FLIGHT_FACTORS.medium!;
  return FLIGHT_FACTORS.long!;
}

function toSuspicious(record: { normalizedCo2eKg: number; activityAmount: number }, threshold = 100000): { isSuspicious: boolean; suspiciousReason: string | undefined } {
  if (record.normalizedCo2eKg > threshold) {
    return { isSuspicious: true, suspiciousReason: `CO₂e value unusually high: ${record.normalizedCo2eKg.toFixed(0)} kg` };
  }
  if (record.activityAmount <= 0) {
    return { isSuspicious: true, suspiciousReason: "Activity amount is zero or negative" };
  }
  return { isSuspicious: false, suspiciousReason: undefined };
}

// ────────────────────────────────────────────────────────────
// SAP Ingest — flat-file tab-delimited IDoc-style export
// Expected columns (case-insensitive, tolerates German variants):
//   MENGE / Menge / quantity, MEINS / unit, MATNR / material,
//   WERKS / plant, BUDAT / posting_date, BUKRS / company_code,
//   optional: BKLAS / material_class
// Scope 1 (fuel combustion), category = fuel type
// ────────────────────────────────────────────────────────────
router.post("/ingest/sap", upload.single("file"), async (req, res): Promise<void> => {
  const clientId = parseInt(req.body.clientId, 10);
  if (!clientId || isNaN(clientId)) {
    res.status(400).json({ error: "clientId is required" });
    return;
  }
  if (!req.file) {
    res.status(400).json({ error: "file is required" });
    return;
  }

  const [client] = await db.select().from(clientsTable).where(eq(clientsTable.id, clientId));
  if (!client) {
    res.status(400).json({ error: "Client not found" });
    return;
  }

  const [batch] = await db.insert(ingestionBatchesTable).values({
    clientId,
    sourceType: "sap",
    filename: req.file.originalname,
    status: "processing",
  }).returning();

  let rowCount = 0;
  let errorCount = 0;
  const errors: string[] = [];

  try {
    const content = req.file.buffer.toString("utf-8");
    const rows: Record<string, string>[] = parse(content, {
      delimiter: ["\t", ";", ","],
      columns: true,
      skip_empty_lines: true,
      trim: true,
      relax_column_count: true,
    });

    const normalize = (key: string) => key.toLowerCase().trim().replace(/\s+/g, "_");

    for (const rawRow of rows) {
      rowCount++;
      const row: Record<string, string> = {};
      for (const [k, v] of Object.entries(rawRow)) row[normalize(k)] = String(v ?? "").trim();

      const quantityRaw = row["menge"] ?? row["quantity"] ?? row["qty"] ?? row["amount"] ?? "";
      const unit = (row["meins"] ?? row["unit"] ?? row["uom"] ?? "L").toLowerCase();
      const material = (row["matnr"] ?? row["material"] ?? row["material_description"] ?? "diesel").toLowerCase();
      const plant = row["werks"] ?? row["plant"] ?? "";
      const dateRaw = row["budat"] ?? row["posting_date"] ?? row["date"] ?? "";
      const companyCode = row["bukrs"] ?? row["company_code"] ?? "";

      const quantity = parseFloat(quantityRaw.replace(",", "."));
      if (isNaN(quantity)) {
        errorCount++;
        errors.push(`Row ${rowCount}: invalid quantity "${quantityRaw}"`);
        continue;
      }

      // Date parsing — SAP uses YYYYMMDD, YYYY-MM-DD, or DD.MM.YYYY
      let activityDate: string;
      try {
        if (/^\d{8}$/.test(dateRaw)) {
          activityDate = `${dateRaw.slice(0, 4)}-${dateRaw.slice(4, 6)}-${dateRaw.slice(6, 8)}`;
        } else if (/^\d{2}\.\d{2}\.\d{4}$/.test(dateRaw)) {
          const [d, m, y] = dateRaw.split(".");
          activityDate = `${y}-${m}-${d}`;
        } else {
          const parsed = new Date(dateRaw);
          activityDate = isNaN(parsed.getTime()) ? new Date().toISOString().slice(0, 10) : parsed.toISOString().slice(0, 10);
        }
      } catch {
        activityDate = new Date().toISOString().slice(0, 10);
      }

      // Determine fuel type and emission factor
      let fuelType = "diesel";
      for (const key of Object.keys(FUEL_FACTORS)) {
        if (material.includes(key)) { fuelType = key; break; }
      }
      const factor = FUEL_FACTORS[fuelType] ?? FUEL_FACTORS["diesel"]!;

      // Unit normalization to litres
      let litres = quantity;
      if (unit === "gal" || unit === "gallon") litres = quantity * 3.78541;
      else if (unit === "m3" || unit === "cbm") litres = quantity * 1000;
      else if (unit === "kg") litres = quantity; // approximate for gases

      const normalizedCo2eKg = litres * factor.kgCo2ePerLitre;
      const suspicious = toSuspicious({ normalizedCo2eKg, activityAmount: quantity });

      await db.insert(emissionsRecordsTable).values({
        batchId: batch.id,
        clientId,
        scope: factor.scope,
        category: fuelType,
        sourceType: "sap",
        rawData: rawRow as Record<string, unknown>,
        activityAmount: litres,
        activityUnit: "litres",
        normalizedCo2eKg,
        activityDate,
        location: plant || companyCode || null,
        description: `${material} combustion — plant ${plant} (${companyCode})`,
        ...suspicious,
      });
    }
  } catch (err) {
    logger.error({ err }, "SAP parse error");
    errorCount++;
    errors.push("Failed to parse file: " + String(err));
  }

  const [updated] = await db
    .update(ingestionBatchesTable)
    .set({
      status: errorCount === rowCount && rowCount > 0 ? "failed" : "completed",
      rowCount,
      errorCount,
      errorDetails: errors.length > 0 ? errors.slice(0, 10).join("\n") : null,
      completedAt: new Date(),
    })
    .where(eq(ingestionBatchesTable.id, batch.id))
    .returning();

  res.status(201).json(updated);
});

// ────────────────────────────────────────────────────────────
// Utility Ingest — portal CSV export from UK/US utility meters
// Expected columns: meter_id, period_start, period_end,
//   consumption_kwh (or consumption + unit), location/site
// Scope 2 (purchased electricity), category = electricity
// ────────────────────────────────────────────────────────────
router.post("/ingest/utility", upload.single("file"), async (req, res): Promise<void> => {
  const clientId = parseInt(req.body.clientId, 10);
  if (!clientId || isNaN(clientId)) {
    res.status(400).json({ error: "clientId is required" });
    return;
  }
  if (!req.file) {
    res.status(400).json({ error: "file is required" });
    return;
  }

  const [client] = await db.select().from(clientsTable).where(eq(clientsTable.id, clientId));
  if (!client) {
    res.status(400).json({ error: "Client not found" });
    return;
  }

  const [batch] = await db.insert(ingestionBatchesTable).values({
    clientId,
    sourceType: "utility",
    filename: req.file.originalname,
    status: "processing",
  }).returning();

  let rowCount = 0;
  let errorCount = 0;
  const errors: string[] = [];

  try {
    const content = req.file.buffer.toString("utf-8");
    const rows: Record<string, string>[] = parse(content, {
      delimiter: [",", ";", "\t"],
      columns: true,
      skip_empty_lines: true,
      trim: true,
      relax_column_count: true,
    });

    const normalize = (key: string) => key.toLowerCase().trim().replace(/[\s-]+/g, "_");

    for (const rawRow of rows) {
      rowCount++;
      const row: Record<string, string> = {};
      for (const [k, v] of Object.entries(rawRow)) row[normalize(k)] = String(v ?? "").trim();

      const kwhRaw = row["consumption_kwh"] ?? row["kwh"] ?? row["usage_kwh"] ?? row["consumption"] ?? "";
      const unit = (row["unit"] ?? row["uom"] ?? "kwh").toLowerCase().replace(/\s/g, "");
      const periodStart = row["period_start"] ?? row["start_date"] ?? row["billing_start"] ?? row["date"] ?? "";
      const meterId = row["meter_id"] ?? row["meter"] ?? row["account_number"] ?? "";
      const site = row["location"] ?? row["site"] ?? row["facility"] ?? row["address"] ?? "";

      let kwh = parseFloat(kwhRaw.replace(",", ""));
      if (isNaN(kwh)) {
        errorCount++;
        errors.push(`Row ${rowCount}: invalid consumption "${kwhRaw}"`);
        continue;
      }

      // Unit conversion: MWh → kWh
      if (unit === "mwh" || unit === "mw") kwh *= 1000;
      else if (unit === "gwh" || unit === "gw") kwh *= 1_000_000;

      let activityDate: string;
      try {
        const d = new Date(periodStart);
        activityDate = isNaN(d.getTime()) ? new Date().toISOString().slice(0, 10) : d.toISOString().slice(0, 10);
      } catch {
        activityDate = new Date().toISOString().slice(0, 10);
      }

      const normalizedCo2eKg = kwh * ELECTRICITY_FACTOR_KG_PER_KWH;
      const suspicious = toSuspicious({ normalizedCo2eKg, activityAmount: kwh });

      await db.insert(emissionsRecordsTable).values({
        batchId: batch.id,
        clientId,
        scope: "scope2",
        category: "electricity",
        sourceType: "utility",
        rawData: rawRow as Record<string, unknown>,
        activityAmount: kwh,
        activityUnit: "kWh",
        normalizedCo2eKg,
        activityDate,
        location: site || meterId || null,
        description: `Electricity — meter ${meterId}${site ? ` at ${site}` : ""}`,
        ...suspicious,
      });
    }
  } catch (err) {
    logger.error({ err }, "Utility parse error");
    errorCount++;
    errors.push("Failed to parse file: " + String(err));
  }

  const [updated] = await db
    .update(ingestionBatchesTable)
    .set({
      status: errorCount === rowCount && rowCount > 0 ? "failed" : "completed",
      rowCount,
      errorCount,
      errorDetails: errors.length > 0 ? errors.slice(0, 10).join("\n") : null,
      completedAt: new Date(),
    })
    .where(eq(ingestionBatchesTable.id, batch.id))
    .returning();

  res.status(201).json(updated);
});

// ────────────────────────────────────────────────────────────
// Travel Ingest — Navan/Concur CSV export
// Expected columns: trip_type (flight/hotel/ground), origin,
//   destination, distance_km, nights, passengers, date,
//   traveler, department
// Scope 3 (business travel), categories: flight/hotel/ground_transport
// ────────────────────────────────────────────────────────────
router.post("/ingest/travel", upload.single("file"), async (req, res): Promise<void> => {
  const clientId = parseInt(req.body.clientId, 10);
  if (!clientId || isNaN(clientId)) {
    res.status(400).json({ error: "clientId is required" });
    return;
  }
  if (!req.file) {
    res.status(400).json({ error: "file is required" });
    return;
  }

  const [client] = await db.select().from(clientsTable).where(eq(clientsTable.id, clientId));
  if (!client) {
    res.status(400).json({ error: "Client not found" });
    return;
  }

  const [batch] = await db.insert(ingestionBatchesTable).values({
    clientId,
    sourceType: "travel",
    filename: req.file.originalname,
    status: "processing",
  }).returning();

  let rowCount = 0;
  let errorCount = 0;
  const errors: string[] = [];

  try {
    const content = req.file.buffer.toString("utf-8");
    const rows: Record<string, string>[] = parse(content, {
      delimiter: [",", ";", "\t"],
      columns: true,
      skip_empty_lines: true,
      trim: true,
      relax_column_count: true,
    });

    const normalize = (key: string) => key.toLowerCase().trim().replace(/[\s-]+/g, "_");

    for (const rawRow of rows) {
      rowCount++;
      const row: Record<string, string> = {};
      for (const [k, v] of Object.entries(rawRow)) row[normalize(k)] = String(v ?? "").trim();

      const tripType = (row["trip_type"] ?? row["type"] ?? row["category"] ?? "flight").toLowerCase();
      const dateRaw = row["date"] ?? row["departure_date"] ?? row["travel_date"] ?? row["check_in"] ?? "";
      const traveler = row["traveler"] ?? row["employee"] ?? row["name"] ?? "";
      const department = row["department"] ?? row["cost_center"] ?? row["team"] ?? "";

      let activityDate: string;
      try {
        const d = new Date(dateRaw);
        activityDate = isNaN(d.getTime()) ? new Date().toISOString().slice(0, 10) : d.toISOString().slice(0, 10);
      } catch {
        activityDate = new Date().toISOString().slice(0, 10);
      }

      let normalizedCo2eKg = 0;
      let activityAmount = 0;
      let activityUnit = "";
      let category = "flight";
      let description = "";

      if (tripType.includes("flight") || tripType.includes("air")) {
        category = "flight";
        const origin = row["origin"] ?? row["departure"] ?? row["from"] ?? "";
        const dest = row["destination"] ?? row["arrival"] ?? row["to"] ?? "";
        const distanceRaw = row["distance_km"] ?? row["distance"] ?? row["km"] ?? "";
        const passengersRaw = row["passengers"] ?? row["pax"] ?? "1";
        const passengers = parseInt(passengersRaw, 10) || 1;

        let distanceKm = parseFloat(distanceRaw);
        if (isNaN(distanceKm) || distanceKm <= 0) {
          distanceKm = getFlightDistance(origin, dest);
        }

        const factor = getFlightFactor(distanceKm);
        normalizedCo2eKg = distanceKm * factor * passengers;
        activityAmount = distanceKm;
        activityUnit = "km";
        description = `Flight ${origin}→${dest} (${distanceKm} km, ${passengers} pax) — ${traveler}${department ? ` [${department}]` : ""}`;
      } else if (tripType.includes("hotel") || tripType.includes("accommodation") || tripType.includes("lodg")) {
        category = "hotel";
        const nightsRaw = row["nights"] ?? row["duration_nights"] ?? row["check_out"] ?? "1";
        const nights = parseFloat(nightsRaw) || 1;
        normalizedCo2eKg = nights * HOTEL_FACTOR_KG_PER_NIGHT;
        activityAmount = nights;
        activityUnit = "nights";
        const location = row["hotel"] ?? row["property"] ?? row["destination"] ?? row["city"] ?? "";
        description = `Hotel ${location} (${nights} nights) — ${traveler}${department ? ` [${department}]` : ""}`;
      } else {
        category = "ground_transport";
        const distanceRaw = row["distance_km"] ?? row["distance"] ?? row["km"] ?? "50";
        const distanceKm = parseFloat(distanceRaw) || 50;
        normalizedCo2eKg = distanceKm * GROUND_FACTOR_KG_PER_KM;
        activityAmount = distanceKm;
        activityUnit = "km";
        const from = row["origin"] ?? row["from"] ?? row["pickup"] ?? "";
        const to = row["destination"] ?? row["to"] ?? row["dropoff"] ?? "";
        description = `Ground transport ${from}→${to} (${distanceKm} km) — ${traveler}${department ? ` [${department}]` : ""}`;
      }

      const suspicious = toSuspicious({ normalizedCo2eKg, activityAmount });

      await db.insert(emissionsRecordsTable).values({
        batchId: batch.id,
        clientId,
        scope: "scope3",
        category,
        sourceType: "travel",
        rawData: rawRow as Record<string, unknown>,
        activityAmount,
        activityUnit,
        normalizedCo2eKg,
        activityDate,
        location: row["origin"] ?? row["city"] ?? null,
        description,
        ...suspicious,
      });
    }
  } catch (err) {
    logger.error({ err }, "Travel parse error");
    errorCount++;
    errors.push("Failed to parse file: " + String(err));
  }

  const [updated] = await db
    .update(ingestionBatchesTable)
    .set({
      status: errorCount === rowCount && rowCount > 0 ? "failed" : "completed",
      rowCount,
      errorCount,
      errorDetails: errors.length > 0 ? errors.slice(0, 10).join("\n") : null,
      completedAt: new Date(),
    })
    .where(eq(ingestionBatchesTable.id, batch.id))
    .returning();

  res.status(201).json(updated);
});

export default router;
